Implement RFC8910: captive portal dhcp options

rpigott opened the pull request:
This adds support for recording the captive portal advertised by dhcp option 114. It can be retrieved from the Describe link json output.

I figured since we are dropping the DVE-2018-0001 kludge now might be a good time to improve captive portal support with DHCP option 114.

~~Doesn't appear to actually work atm though... the portal is not recorded in our state file despite option 114 appearing in the dhcp option list according to tcpdump.~~

poettering commented:
This probably should declare that this is an implementation of RFC 8910, i.e. https://www.rfc-editor.org/rfc/rfc8910.html

poettering commented on src/network/networkd.c:
```diff
@@ -1,5 +1,3 @@
-/* SPDX-License-Identifier: LGPL-2.1-or-later */
-
```
uh?

rpigott replied:
idk how that happened.

poettering reviewed (changes requested):
would be good to implement the same for dhcp6 (which uses a different option nr though) at the same time. and the IPv6RA support would be nice too.

poettering commented:
also needs man page updates

rpigott commented:
> would be good to implement the same for dhcp6 (which uses a different option nr though) at the same time. and the IPv6RA support would be nice too.

Yeah I plan to add these.

rpigott commented:
Ok the dhcp4 option actually works as intended now.

> This probably should declare that this is an implementation of RFC 8910, i.e. https://www.rfc-editor.org/rfc/rfc8910.html

@poettering Sounds good, but how do you mean? Just in the commit message?

RFC 8910 also includes this idea where a portal urn that is exactly "urn:ietf:params:capport:unrestricted" means there is explicitly no captive portal, and clients can skip any kind of detection. Should we have some kind of captive portal state in the api or just let clients figure that one out?

rpigott commented:
Okay, added dhcpv6 + ipv6ra options and updated systemd.network(5) to include all the UseCaptivePortal directives. Should be in a reviewable state now.

Interesting questions:

1. Do we care if the network advertises different portals using some of the 3 options? rfc says they should not do that.
2. Should we have some state to record whether or not the advertised portal is the special value "urn:ietf:params:capport:unrestricted"?
3. Is it fine just having it in the Describe dump or maybe there should be a property that changes?

yuwata commented on man/systemd.network.xml:
```diff
@@ -1969,6 +1969,14 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
+        <term><varname>UseCaptivePortal=</varname></term>
+        <listitem>
+          <para>When true (the default), the captive portal advertised by the DHCP server will be recorded
+          and made available to client programs.</para>
```
Maybe, better to explicitly mention `networkctl`.

rpigott replied:
Ok, done.

yuwata commented on src/network/networkd-dhcp4.c:
```diff
@@ -1426,6 +1426,12 @@ static int dhcp4_configure(Link *link) {
+                        if (r < 0)
+                                return log_link_debug_errno(link, r, "DHCPv4 CLIENT: Failed to set request flag for captive portal: %m");
+                }
+
```
doubled empty line.

yuwata commented on src/network/networkd-json.c:
```diff
@@ -873,6 +873,27 @@ static int sip_build_json(Link *link, JsonVariant **ret) {
+        }
+
+        if (link->network->dhcp_use_captive_portal) {
+                r = sd_dhcp_lease_get_captive_portal(link->dhcp_lease, &captive_portal);
+                if (r < 0 && r != -ENODATA)
+                        return r;
```
Missing `r == -ENODATA` case handling?

rpigott replied:
Well here I don't really care about the ENODATA case I think. These functions return ENODATA to match the similar getters. In the ENODATA case captive_portal will remain NULL and the dhcp6 and ipv6ra blocks will have a chance to fill it.

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -264,6 +271,8 @@ int manager_save(Manager *m) {
         online_state_str = link_online_state_to_string(online_state);
         if (online_state_str)
                 fprintf(f, "ONLINE_STATE=%s\n", online_state_str);
+        if (captive_portal)
+                fprintf(f, "CAPTIVE_PORTAL=%s\n", captive_portal);
```
Why saved in the manager state??
That may be OK, but please also save it in link state (or, may be in dhcp lease file??).
Otherwise, sd_network_link_get_captive_portal() will never work.

rpigott replied:
I didn't know where to put it honestly. I'll look at putting it in the lease state, that sounds correct.

rpigott replied:
I settled on putting it in the link state.

yuwata commented on src/libsystemd-network/sd-dhcp6-lease.c:
```diff
@@ -445,6 +445,37 @@ int sd_dhcp6_lease_get_fqdn(sd_dhcp6_lease *lease, const char **ret) {
+        assert(lease);
+        assert(optval || optlen == 0);
+
+        mfree(lease->captive_portal);
```
Please do not free the previous result earlier, but use `free_and_replace()` at the end.

yuwata commented on src/libsystemd-network/sd-dhcp6-lease.c:
```diff
@@ -445,6 +445,37 @@ int sd_dhcp6_lease_get_fqdn(sd_dhcp6_lease *lease, const char **ret) {
+        assert(optval || optlen == 0);
+
+        mfree(lease->captive_portal);
+        if (optlen == 0) {
```
Unnecessary `{}`.

yuwata commented on src/libsystemd-network/ndisc-router.c:
```diff
@@ -715,3 +715,29 @@ int sd_ndisc_router_dnssl_get_lifetime(sd_ndisc_router *rt, uint32_t *ret_sec) {
         return 0;
 }
+
+int sd_ndisc_router_captive_portal_get_uri(sd_ndisc_router *rt, const char **uri, size_t *size) {
```
uri -> ret_uri
size -> ret_size

yuwata commented on src/libsystemd-network/ndisc-router.c:
```diff
@@ -715,3 +715,29 @@ int sd_ndisc_router_dnssl_get_lifetime(sd_ndisc_router *rt, uint32_t *ret_sec) {
+                return -EBADMSG;
+
+        *uri = nd_opt_captive_portal + 2;
+        *size = strnlen(*uri, length);
```
length -> length - 2

Maybe, better to return `NULL` when `length == 2` ??

rpigott replied:
> length -> length - 2

Ah, yep.

> Maybe, better to return NULL when length == 2 ??

Yeah, probably. The dhcp4/6 cases try not to record empty strings at least, which I chose based on the existing behavior of `dhcp_option_parse_string()`. I will make all three options uniformly return NULL in case of empty strings.

yuwata commented on src/network/networkd-json.c:
```diff
@@ -897,6 +897,10 @@ static int captive_portal_build_json(Link *link, JsonVariant **ret) {
                         return r;
         }
 
+        if (link->network->ipv6_accept_ra_use_captive_portal && !captive_portal) {
```
Unnecessary `{}`.

yuwata commented on src/network/networkd-ndisc.c:
```diff
@@ -716,6 +716,37 @@ DEFINE_PRIVATE_HASH_OPS_WITH_KEY_DESTRUCTOR(
+        if (r < 0)
+                return r;
+
+        mfree(link->ndisc_captive_portal);
```
Please use `free_and_replace()` at the end.

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -213,6 +213,10 @@ int manager_save(Manager *m) {
                                 return r;
                 }
 
+                if (!captive_portal && link->network->ipv6_accept_ra_use_captive_portal) {
```
Unnecessary `{}`

yuwata commented on src/libsystemd-network/ndisc-router.c:
```diff
@@ -715,3 +715,29 @@ int sd_ndisc_router_dnssl_get_lifetime(sd_ndisc_router *rt, uint32_t *ret_sec) {
+        if (r == 0)
+                return -EMEDIUMTYPE;
+
+        r = sd_ndisc_router_option_get_raw(rt, (void *)&nd_opt_captive_portal, &length);
+        if (r < 0)
+                return r;
+        if (length < 2)
+                return -EBADMSG;
```
From the rfc https://www.rfc-editor.org/rfc/rfc8910.html#section-2.3-3.1.1.6
> The URI for the captive portal API endpoint to which the user should connect. This MUST be padded with NUL (0x00) to make the total option length (including the Type and Length fields) a multiple of 8 bytes.

So, maybe better to check the length in more detail?

rpigott replied:
It is guaranteed here. The length field in the RA is in units of 8 octets and the  `length` parameter returned in `sd_ndisc_router_option_get_raw()` is exactly the length field \* 8, so length is always a multiple of 8.

The only question is whether or not the client padded the remaining 0-7 bytes with NULs. I use `strnlen` below to find the exact length of the uri without trailing NULs. Should I check something more?

rpigott replied:
Hm I should probably check that the strnlen result doesn't truncate the option before the last 8 bytes.

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -200,6 +201,22 @@ int manager_save(Manager *m) {
                                 return r;
                 }
 
+                if (link->network->dhcp_use_captive_portal) {
+                        r = sd_dhcp_lease_get_captive_portal(link->dhcp_lease, &captive_portal);
+                        if (r < 0 && r != -ENODATA)
+                                return r;
+                }
+
+                if (!captive_portal && link->network->dhcp6_use_captive_portal) {
+                        r = sd_dhcp6_lease_get_captive_portal(link->dhcp6_lease, &captive_portal);
+                        if (r < 0 && r != -ENODATA)
+                                return r;
+                }
+
+                if (!captive_portal && link->network->ipv6_accept_ra_use_captive_portal) {
+                        captive_portal = link->ndisc_captive_portal;
+                }
```
See https://www.rfc-editor.org/rfc/rfc8910.html#name-precedence-of-api-uris.
Please check captive portals obtained through multiple protocols are identical, and log that captive portal provided by DHCPv4 has higher precedence if they are not identical.

rpigott replied:
Done.

yuwata commented on src/network/networkctl.c:
```diff
@@ -1706,6 +1706,7 @@ static int link_status_one(
         (void) sd_network_link_get_route_domains(info->ifindex, &route_domains);
         (void) sd_network_link_get_ntp(info->ifindex, &ntp);
         (void) sd_network_link_get_sip(info->ifindex, &sip);
+        (void) sd_network_link_get_captive_portal(info->ifindex, &captive_portal);
```
Not used??

rpigott replied:
Added a "Captive Portal" field to the `networkctl status <link>` output, so now this is used.

rpigott replied:
```
$ networkctl status mv-enp4s0
[...]
     Activation Policy: up
   Required For Online: yes
        Captive Portal: http://example.com
       DHCP4 Client ID: IAID:0xc3784450/DUID
     DHCP6 Client DUID: DUID-EN/Vendor:0000ab11f29abad7f3b60b5d0000
```

yuwata commented:
Please add a simple testcase in test/test-network/systemd-networkd-tests.py.

rpigott commented:
Added a simple test case.

poettering commented on src/network/networkd-state-file.c:
```diff
@@ -606,6 +608,17 @@ int link_save(Link *link) {
+                }
+
+                if (dhcp_captive_portal)
+                        fprintf(f, "CAPTIVE_PORTAL=%s\n", dhcp_captive_portal);
```
should probably use cescape() to ensure any embedded newlines in the URLs don't confuse things

poettering commented on src/network/networkctl.c:
```diff
@@ -2358,6 +2359,9 @@ static int link_status_one(
         }
 
+        if (captive_portal)
+                table_add_string_line(table, "Captive Portal:", captive_portal);
```
hmm, this should probably escape the field before display, so that embedded control characters/ANSI sequences or so don't fuck things up.

i.e. maybe just call cescape() or so before showing this

poettering commented on src/network/networkd-state-file.c:
```diff
@@ -106,7 +106,8 @@ static int ordered_set_put_in4_addrv(
         _cleanup_ordered_set_free_ OrderedSet *dns = NULL, *ntp = NULL, *sip = NULL, *search_domains = NULL, *route_domains = NULL;
-        const char *operstate_str, *carrier_state_str, *address_state_str, *ipv4_address_state_str, *ipv6_address_state_str, *online_state_str;
+        const char *operstate_str, *carrier_state_str, *address_state_str, *ipv4_address_state_str, *ipv6_address_state_str,
+              *online_state_str;
```
mm?

rpigott replied:
ah woops, leftover from rebase.

poettering commented on src/libsystemd/sd-network/sd-network.c:
```diff
@@ -258,6 +258,10 @@ int sd_network_link_get_sip(int ifindex, char ***ret) {
 }
 
+int sd_network_link_get_captive_portal(int ifindex, char **ret) {
+        return network_link_get_string(ifindex, "CAPTIVE_PORTAL", ret);
```
should unescape i guess

poettering replied:
i.e. cunescape()

poettering commented on src/libsystemd-network/sd-dhcp6-lease.c:
```diff
@@ -445,6 +445,37 @@ int sd_dhcp6_lease_get_fqdn(sd_dhcp6_lease *lease, const char **ret) {
+        assert(optval || optlen == 0);
+
+        if (optlen == 0) {
+                mfree(lease->captive_portal);
```
```c
lease->captive_portal = mfree(lease->captive_portal);
```

poettering commented on src/libsystemd-network/sd-dhcp6-lease.c:
```diff
@@ -445,6 +445,37 @@ int sd_dhcp6_lease_get_fqdn(sd_dhcp6_lease *lease, const char **ret) {
+                return r;
+
+        free_and_replace(lease->captive_portal, captive_portal);
+        return 0;
```
why not use dhcp_option_parse_string() here?

rpigott replied:
My intention was to refuse a trailing NUL byte in this case, as I don't think the comment in dhcp_option_parse_string applies to dhcp6? Not sure though. Even if it doesn't might be prudent to create a dhcp6_option_parse_string...

rpigott replied:
I've gone ahead and added such a function.

poettering commented on src/network/networkd-state-file.c:
```diff
@@ -625,10 +625,22 @@ int link_save(Link *link) {
+                        if (dhcp_captive_portal && strcmp(dhcp_captive_portal, link->ndisc_captive_portal) != 0)
+                                log_link_warning(link, "IPv6RA captive portal (%s) does not match DHCPv4 (%s). Ignorning.",
+                                                link->ndisc_captive_portal, dhcp_captive_portal);
+                        if (dhcp6_captive_portal && strcmp(dhcp6_captive_portal, link->ndisc_captive_portal) != 0)
```
as above, streq()

poettering commented on src/network/networkd-state-file.c:
```diff
@@ -625,10 +625,22 @@ int link_save(Link *link) {
+                else if (link->ndisc_captive_portal)
+                        fprintf(f, "CAPTIVE_PORTAL=%s\n", link->ndisc_captive_portal);
+
 
```
spurious double empty line

poettering commented on src/network/networkd-state-file.c:
```diff
@@ -625,10 +625,22 @@ int link_save(Link *link) {
                 else if (dhcp6_captive_portal)
                         fprintf(f, "CAPTIVE_PORTAL=%s\n", dhcp6_captive_portal);
+                else if (link->ndisc_captive_portal)
+                        fprintf(f, "CAPTIVE_PORTAL=%s\n", link->ndisc_captive_portal);
```
escaping...

poettering commented:
looks good in general. but it's key to properly escape things when writing it to the env file and when displaying it to the user. it's good to pass it around internally as it is, but we need to make sure we don't spit out attacker-controlled data in unsafe ways onto the terminal or any file formats that can't deal with it.

(note the json code will escape internally automatically, so it's safe to pass the data to the json encoder as it is)

rpigott commented:
`cescape` strategy was a bust, as parse_env_file just mangles the escaped values when reading them back.

I've decided to instead do some basic validation of the URI, since it seems pointless to try to preserve all these characters that aren't valid in a URI anyway. So now it will reject/ignore advertised portal uris that contain invalid characters, and we can safely store/print that value. PTAL and let me know what you think of this approach.

rpigott commented:
tumbleweed failure appears to be https://bugzilla.opensuse.org/show_bug.cgi?id=1212674

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -623,10 +623,21 @@ int link_save(Link *link) {
+                if (link->network->ipv6_accept_ra_use_captive_portal && link->ndisc_captive_portal) {
+                        if (dhcp_captive_portal && !streq(dhcp_captive_portal, link->ndisc_captive_portal))
+                                log_link_warning(link, "IPv6RA captive portal (%s) does not match DHCPv4 (%s). Ignorning IPv6RA portal.",
+                                                link->ndisc_captive_portal, dhcp_captive_portal);
```
indentation.

yuwata commented on src/network/networkd-state-file.c:
```diff
@@ -623,10 +623,21 @@ int link_save(Link *link) {
+                                                link->ndisc_captive_portal, dhcp_captive_portal);
+                        if (dhcp6_captive_portal && !streq(dhcp6_captive_portal, link->ndisc_captive_portal))
+                                log_link_warning(link, "IPv6RA captive portal (%s) does not match DHCPv6 (%s). Ignorning IPv6RA portal.",
+                                                link->ndisc_captive_portal, dhcp6_captive_portal);
```
Also here.

yuwata reviewed (changes requested):
Could you rearrange the commits?
1. update sd-dhcp-lease.[ch] and friends, but leave networkd unchanged.
2. update sd-dhcp6-lease.[ch] and freinds, but leave networkd unchanged.
3. update ndisc-router.[ch] and friends, but leave networkd unchanged.
4. add three .network options about the captive portal, and update networkd-state-file.c.
5. update networkd-json.c.
6. update networkctl.c.
7. add test.

rpigott commented on src/network/networkd-ndisc.c:
```diff
@@ -716,6 +716,40 @@ DEFINE_PRIVATE_HASH_OPS_WITH_KEY_DESTRUCTOR(
+        if (r < 0)
+                return r;
+
+        if (len == 0) {
+                mfree(link->ndisc_captive_portal);
+                return 0;
+        }
+
+        r = make_cstring(uri, len, MAKE_CSTRING_REFUSE_TRAILING_NUL, &captive_portal);
+        if (r < 0)
+                return r;
+
+        if (!in_charset(captive_portal, URI_VALID))
+                return -EINVAL;
```
I actually tried this but thought it complicated the code a bit much. `in_charset` only works on proper c strings as it uses `strspn` and there is no `strnspn`. So I added a `in_charset_safe` to operate on the sized ref returned by `sd_ndisc_router_captive_portal` but ultimately just decided to order this check after `make_cstring`, which is similar to the logic in the other captive_portal option parsers.

rpigott commented on src/libsystemd-network/ndisc-router.c:
```diff
@@ -715,3 +715,40 @@ int sd_ndisc_router_dnssl_get_lifetime(sd_ndisc_router *rt, uint32_t *ret_sec) {
+                return 0;
+        }
+
+        /* Check that the message is not truncated by an embedded NUL.
+         * NUL padding to a multiple of 8 is expected. */
+        size_t size = strnlen(nd_opt_captive_portal + 2, length - 2);
+        if (size + 2 != length && ((size + 2) / 8) + 1 != length / 8)
+                return -EBADMSG;
```
> `if (length % 8 != 0)`

`length % 8 != 0 ` is guaranteed here, as it is just the return parameter from `sd_ndisc_router_option_get_raw`, which returns the length field * 8. I will try to write this more clearly. DIV_ROUND_UP is a good suggestion.

rpigott replied:
I've updated it similar to the suggestion. Please check the new logic here.

rpigott commented:
> Could you rearrange the commits?

Here is a version with the requested history. Wasn't sure if the options should still be separate, lmk if I should squash them. Also caught that ipv6ra option default had not been set.

rpigott commented:
fixed whitespace mismatch in the test.

yuwata reviewed (approved):
LGTM. Thank you!

evverx commented:
There is a double-free somewhere: https://github.com/systemd/systemd/issues/28229

evverx commented:
As far as I can see that part isn't covered by any tests at all. It would probably make sense to cover it too just in case.
